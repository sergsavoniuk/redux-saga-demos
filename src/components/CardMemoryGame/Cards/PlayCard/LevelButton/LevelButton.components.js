import styled from 'styled-components';

export const Button = styled.button`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  padding: 0;
  border: none;
  color: #fff;
  font-family: 'Open Sans';
  font-size: 0.6em;
  font-weight: bold;
  background-color: transparent;
  text-transform: uppercase;
  text-decoration: none;

  &:hover {
    background-color: #ffffff1a;
  }
`;
