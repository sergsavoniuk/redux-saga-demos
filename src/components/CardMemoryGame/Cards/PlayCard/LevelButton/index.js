export { default, LevelButton } from './LevelButton';
