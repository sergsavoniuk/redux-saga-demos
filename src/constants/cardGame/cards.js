export const CARDS = {
  Figures: 'Figures',
  Mock: 'Mock',
  Instructions: 'Instructions',
  Play: 'Play',
};
