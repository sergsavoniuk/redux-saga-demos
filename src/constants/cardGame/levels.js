export const LEVELS = {
  Casual: 'casual',
  Medium: 'medium',
  Hard: 'hard',
};
