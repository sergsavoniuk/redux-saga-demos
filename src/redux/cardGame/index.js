export { default, ActionTypes, ActionCreators, Selectors } from './cardGame';
