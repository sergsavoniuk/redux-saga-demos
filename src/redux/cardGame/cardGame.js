import { createSelector } from 'reselect';

import { GAME_STATUSES } from 'constants/cardGame/statuses';

// Action Types
export const ActionTypes = {
  CHOOSE_LEVEL: '@cardGame/CHOOSE_LEVEL',
  SET_STATUS: '@card/SET_STATUS',
  FILL_CARDS: '@cardGame/FILL_CARDS',
  FLIP_CARD: '@cardGame/FLIP_CARD',
  CHECK_FLIPPED_CARDS: '@cardName/CHECK_FLIPPED_CARDS',
  FINISH_GAME: '@cardGame/FINISH_GAME',
  UPDATE_FLIPS_STATISTICS: '@cardGame/UPDATE_FLIPS_STATISTICS',
  UPDATE_FIGURES_STATISTICS: '@cardGame/UPDATE_FIGURES_STATISTICS',
  UPDATE_BEST_TIME_STATISTICS: '@cardGame/UPDATE_BEST_TIME_STATISTICS',
};

// Action Creators
export const ActionCreators = {
  chooseLevel(level) {
    return {
      type: ActionTypes.CHOOSE_LEVEL,
      payload: {
        level,
      },
    };
  },

  setStatus(status) {
    return {
      type: ActionTypes.SET_STATUS,
      payload: {
        status,
      },
    };
  },

  fillCards(cards) {
    return {
      type: ActionTypes.FILL_CARDS,
      payload: {
        cards,
      },
    };
  },

  flipCard(cardId) {
    return {
      type: ActionTypes.FLIP_CARD,
      payload: {
        cardId,
      },
    };
  },

  checkFlippedCards() {
    return {
      type: ActionTypes.CHECK_FLIPPED_CARDS,
    };
  },

  finishGame() {
    return {
      type: ActionTypes.FINISH_GAME,
    };
  },

  updateFlipsStatistics(flippedCardsIds) {
    return {
      type: ActionTypes.UPDATE_FLIPS_STATISTICS,
      payload: {
        flippedCardsIds,
      },
    };
  },

  updateFiguresStatistics(key) {
    return {
      type: ActionTypes.UPDATE_FIGURES_STATISTICS,
      payload: {
        key,
      },
    };
  },

  updateBestTimeStatistics({ key, time }) {
    return {
      type: ActionTypes.UPDATE_BEST_TIME_STATISTICS,
      payload: {
        key,
        time,
      },
    };
  },
};

const initialState = {
  cards: {},
  flippedCardsIds: [],
  status: GAME_STATUSES.Pending,
  level: null,
  statistics: {
    won: 0,
    lost: 0,
    matchedFlips: 0,
    wrongFlips: 0,
    casualBestTime: null,
    mediumBestTime: null,
    hardBestTime: null,
  },
};

// Reducer
export default function cardGameReducer(state = initialState, action) {
  switch (action.type) {
    case ActionTypes.CHOOSE_LEVEL: {
      return {
        ...state,
        level: action.payload.level,
      };
    }
    case ActionTypes.SET_STATUS: {
      return {
        ...state,
        status: action.payload.status,
      };
    }
    case ActionTypes.FILL_CARDS: {
      return {
        ...state,
        cards: action.payload.cards,
        flippedCardsIds: [],
      };
    }
    case ActionTypes.FLIP_CARD: {
      const { cardId } = action.payload;
      const flippedCardsIds = state.flippedCardsIds;
      return {
        ...state,
        flippedCardsIds: flippedCardsIds.includes(cardId)
          ? flippedCardsIds
          : flippedCardsIds.concat(cardId),
      };
    }
    case ActionTypes.CHECK_FLIPPED_CARDS: {
      const [firstCardId, secondCardId] = state.flippedCardsIds;
      const flippedCardsIdsMatched =
        state.cards[firstCardId].key === state.cards[secondCardId].key;
      return {
        ...state,
        flippedCardsIds: [],
        cards: flippedCardsIdsMatched
          ? {
              ...state.cards,
              [firstCardId]: {
                ...state.cards[firstCardId],
                isGuessed: true,
              },
              [secondCardId]: {
                ...state.cards[secondCardId],
                isGuessed: true,
              },
            }
          : state.cards,
      };
    }
    case ActionTypes.UPDATE_FLIPS_STATISTICS: {
      const [firstCardId, secondCardId] = action.payload.flippedCardsIds;
      const flippedCardsIdsMatched =
        state.cards[firstCardId].key === state.cards[secondCardId].key;
      return {
        ...state,
        statistics: {
          ...state.statistics,
          matchedFlips: flippedCardsIdsMatched
            ? state.statistics.matchedFlips + 1
            : state.statistics.matchedFlips,
          wrongFlips: flippedCardsIdsMatched
            ? state.statistics.wrongFlips
            : state.statistics.wrongFlips + 1,
        },
      };
    }
    case ActionTypes.UPDATE_FIGURES_STATISTICS: {
      const { key } = action.payload;
      return {
        ...state,
        statistics: {
          ...state.statistics,
          [key]: state.statistics[key] + 1,
        },
      };
    }
    case ActionTypes.UPDATE_BEST_TIME_STATISTICS: {
      const { key, time } = action.payload;
      return {
        ...state,
        statistics: {
          ...state.statistics,
          [key]:
            state.statistics[key] !== null && state.statistics[key] < time
              ? state.statistics[key]
              : time,
        },
      };
    }
    default:
      return state;
  }
}

// Selectors
export const Selectors = {
  getCardIds(state) {
    return Object.keys(state.cardGame.cards);
  },

  getCardById(state, cardId) {
    return state.cardGame.cards[cardId];
  },

  getFlippedCardsIds(state) {
    return state.cardGame.flippedCardsIds;
  },

  getStatistics(state) {
    return state.cardGame.statistics;
  },

  unguessedCardsCountSelector: createSelector(
    state => state.cardGame.cards,
    cards =>
      Object.keys(cards).reduce((acc, card) => {
        if (!cards[card].isGuessed) {
          acc.push(card);
        }
        return acc;
      }, []).length,
  ),
};
