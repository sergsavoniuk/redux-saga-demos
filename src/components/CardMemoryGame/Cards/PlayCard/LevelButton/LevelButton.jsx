import React from 'react';
import { string, func } from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { Button } from './LevelButton.components';
import { ActionCreators } from 'redux/cardGame';
import { ROUTES } from 'constants/routes';

export function LevelButton({ level, chooseLevel }) {
  function handleChooseLevel() {
    chooseLevel(level);
  }

  return (
    <Button
      as={Link}
      to={{
        pathname: `${ROUTES.CardGameApp.Play}`,
        search: `?level=${level}`,
      }}
      onClick={handleChooseLevel}
    >
      {level}
    </Button>
  );
}

LevelButton.propTypes = {
  level: string.isRequired,
  chooseLevel: func.isRequired,
};

export default connect(
  null,
  {
    chooseLevel: ActionCreators.chooseLevel,
  },
)(LevelButton);
