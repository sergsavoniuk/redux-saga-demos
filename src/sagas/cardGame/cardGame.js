import { take, put, all, race, fork, delay } from 'redux-saga/effects';
import { goBack, LOCATION_CHANGE } from 'connected-react-router';

import { ActionTypes, ActionCreators } from 'redux/cardGame';
import { GAME_STATUSES } from 'constants/cardGame/statuses';
import { LEVELS } from 'constants/cardGame/levels';
import { LEVEL_TO_TIME } from 'constants/cardGame/levelToTime';
import { GAME_RESULTS } from 'constants/cardGame/gameResults';
import { ROUTES } from 'constants/routes';

const { Casual, Medium, Hard } = LEVELS;
const { Won, Lost, Abandoned } = GAME_RESULTS;

const LEVEL_TO_BOARD_CELLS = {
  [Casual]: 16,
  [Medium]: 20,
  [Hard]: 32,
};

function shuffle(a) {
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

function generateAndFillArray(length) {
  const array = [];
  for (let i = 1; i <= length; ++i) {
    array.push(i);
  }
  return array;
}

export function generateCards(level) {
  const cardArray = shuffle([
    ...generateAndFillArray(LEVEL_TO_BOARD_CELLS[level] / 2),
    ...generateAndFillArray(LEVEL_TO_BOARD_CELLS[level] / 2),
  ]);

  const cards = {};
  cardArray.forEach((card, index) => {
    cards[index] = {
      key: card,
      content: card,
      isGuessed: false,
    };
  });
  return cards;
}

export function* cardGameWorkerSaga(gameLevel) {
  const levelTime = LEVEL_TO_TIME[gameLevel];

  let timeSpent = 0;

  while (true) {
    const startTime = performance.now();

    const { finish, changeLocation, timeout, pause } = yield race({
      finish: take(ActionTypes.FINISH_GAME),
      changeLocation: take(LOCATION_CHANGE),
      timeout: delay(
        Math.abs(levelTime - timeSpent) < 1000
          ? levelTime
          : levelTime - timeSpent,
      ),
      pause: take(ActionTypes.SET_STATUS),
    });

    timeSpent = performance.now() - startTime;

    const actions = [
      ActionCreators.setStatus(GAME_STATUSES.Finished),
      goBack(),
    ];

    if (timeout) {
      yield all(
        actions
          .concat(ActionCreators.updateFiguresStatistics(Lost))
          .map(action => put(action)),
      );
      break;
    } else if (
      finish ||
      (changeLocation &&
        changeLocation.payload.location.pathname !== ROUTES.CardGameApp.Play)
    ) {
      if (changeLocation || finish.payload.abandoned) {
        const appliedAction = finish ? actions : actions.slice(0, 1);
        yield all(
          appliedAction
            .concat(ActionCreators.updateFiguresStatistics(Abandoned))
            .map(action => put(action)),
        );
      } else {
        yield all(
          actions
            .concat([
              ActionCreators.updateFiguresStatistics(Won),
              ActionCreators.updateBestTimeStatistics({
                key: `${gameLevel}BestTime`,
                time: timeSpent,
              }),
            ])
            .map(action => put(action)),
        );
      }
      break;
    } else if (pause) {
      const { finishGame } = yield race({
        setRunningStatus: take(ActionTypes.SET_STATUS),
        finishGame: take(ActionTypes.FINISH_GAME),
      });

      if (finishGame) {
        yield all(
          actions
            .concat(ActionCreators.updateFiguresStatistics(Abandoned))
            .map(action => put(action)),
        );
        break;
      }
    }
  }
}

export default function* cardGameWatcher() {
  while (true) {
    const {
      payload: { level },
    } = yield take(ActionTypes.CHOOSE_LEVEL);

    yield all([
      put(ActionCreators.setStatus(GAME_STATUSES.Running)),
      put(ActionCreators.fillCards(generateCards(level))),
    ]);

    yield fork(cardGameWorkerSaga, level);
  }
}
