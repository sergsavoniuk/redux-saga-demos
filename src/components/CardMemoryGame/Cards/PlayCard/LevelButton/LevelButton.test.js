import React from 'react';
import { shallow } from 'enzyme';

import { LevelButton } from './LevelButton';
import { LEVELS } from 'constants/cardGame/levels';

const { Casual, Medium } = LEVELS;

describe('Testing LevelButton component', () => {
  let props;

  beforeEach(() => {
    props = {
      level: Casual,
      chooseLevel: jest.fn(),
    };
  });

  test('renders correctly', () => {
    const wrapper = shallow(<LevelButton {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('should change a game level after a click', () => {
    const wrapper = shallow(<LevelButton {...props} />);

    expect(wrapper.text()).toMatch(Casual);

    wrapper.simulate('click');
    wrapper.setProps({ level: Medium });

    expect(props.chooseLevel).toHaveBeenCalled();
    expect(wrapper.text()).toMatch(Medium);
    expect(wrapper).toMatchSnapshot();
  });
});
