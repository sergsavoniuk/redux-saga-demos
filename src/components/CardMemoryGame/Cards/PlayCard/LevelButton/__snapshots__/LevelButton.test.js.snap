// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing LevelButton component renders correctly 1`] = `
<styled.button
  as={[Function]}
  onClick={[Function]}
  to={
    Object {
      "pathname": "/apps/card-memory-game/play",
      "search": "?level=casual",
    }
  }
>
  casual
</styled.button>
`;

exports[`Testing LevelButton component should change a game level after a click 1`] = `
<styled.button
  as={[Function]}
  onClick={[Function]}
  to={
    Object {
      "pathname": "/apps/card-memory-game/play",
      "search": "?level=medium",
    }
  }
>
  medium
</styled.button>
`;
